### ISSUE TITLE
macOS Universal build verification is broken

### ISSUE THREAD

**Issue by capital-G** [supercollider/supercollider#6352]
Once again, (most likely) a brew update breaks our CI :/

See https://github.com/supercollider/supercollider/actions/runs/9501252970/job/26192773490

```log
2024-06-13T16:22:25.7103150Z -- Verifying app
2024-06-13T16:22:25.7165060Z -- ===========================================================================
2024-06-13T16:22:25.7166950Z -- Analyzing app='/Users/runner/work/supercollider/supercollider/build/Install/SuperCollider/SuperCollider.app'
2024-06-13T16:22:25.7169180Z -- bundle='/Users/runner/work/supercollider/supercollider/build/Install/SuperCollider/SuperCollider.app'
2024-06-13T16:22:25.7171520Z -- executable='/Users/runner/work/supercollider/supercollider/build/Install/SuperCollider/SuperCollider.app/Contents/MacOS/SuperCollider'
2024-06-13T16:22:25.7173060Z -- valid='1'
2024-06-13T16:22:26.2895860Z -- executable file 1: /Users/runner/work/supercollider/supercollider/build/Install/SuperCollider/SuperCollider.app/Contents/MacOS/SuperCollider
2024-06-13T16:22:27.1989680Z -- 
2024-06-13T16:22:27.1996390Z warning: cannot resolve item '/usr/local/Cellar/glib/2.80.3/lib/libglib-2.0.0.dylib'
2024-06-13T16:22:27.1997310Z 
2024-06-13T16:22:27.1997540Z   possible problems:
2024-06-13T16:22:27.1998170Z     need more directories?
2024-06-13T16:22:27.2108150Z     need to use InstallRequiredSystemLibraries?
2024-06-13T16:22:27.2109000Z     run in install tree instead of build tree?
2024-06-13T16:22:27.2111680Z 
2024-06-13T16:22:27.7322430Z -- executable file 2: /Users/runner/work/supercollider/supercollider/build/Install/SuperCollider/SuperCollider.app/Contents/MacOS/sclang
2024-06-13T16:22:27.9222680Z -- 
2024-06-13T16:22:27.9224360Z warning: cannot resolve item '/usr/local/Cellar/glib/2.80.3/lib/libglib-2.0.0.dylib'
2024-06-13T16:22:27.9227740Z 
2024-06-13T16:22:27.9230180Z   possible problems:
2024-06-13T16:22:27.9231320Z     need more directories?
2024-06-13T16:22:27.9231880Z     need to use InstallRequiredSystemLibraries?
2024-06-13T16:22:27.9233030Z     run in install tree instead of build tree?
2024-06-13T16:22:27.9234290Z 
2024-06-13T16:22:28.4955530Z -- executable file 3: /Users/runner/work/supercollider/supercollider/build/Install/SuperCollider/SuperCollider.app/Contents/Resources/supernova
2024-06-13T16:22:28.6467340Z -- executable file 4: /Users/runner/work/supercollider/supercollider/build/Install/SuperCollider/SuperCollider.app/Contents/Resources/scsynth
2024-06-13T16:22:28.7635280Z -- executable file 5: /Users/runner/work/supercollider/supercollider/build/Install/SuperCollider/SuperCollider.app/Contents/Frameworks/QtWebEngineCore.framework/Versions/5/Helpers/QtWebEngineProcess.app/Contents/MacOS/QtWebEngineProcess
2024-06-13T16:22:28.9590680Z -- 
2024-06-13T16:22:28.9592510Z warning: cannot resolve item '/usr/local/Cellar/glib/2.80.3/lib/libglib-2.0.0.dylib'
2024-06-13T16:22:28.9603980Z 
2024-06-13T16:22:28.9604370Z   possible problems:
2024-06-13T16:22:28.9605000Z     need more directories?
2024-06-13T16:22:28.9605550Z     need to use InstallRequiredSystemLibraries?
2024-06-13T16:22:28.9606590Z     run in install tree instead of build tree?
2024-06-13T16:22:28.9607210Z 
2024-06-13T16:22:29.2861570Z -- verified='0'
2024-06-13T16:22:29.2876320Z -- info='external prerequisites found:
2024-06-13T16:22:29.2881550Z f='/Users/runner/work/supercollider/supercollider/build/Install/SuperCollider/SuperCollider.app/Contents/MacOS/SuperCollider'
2024-06-13T16:22:29.2889540Z external_prereqs='/usr/local/Cellar/glib/2.80.3/lib/libglib-2.0.0.dylib'
2024-06-13T16:22:29.2891790Z ;external prerequisites found:
2024-06-13T16:22:29.2897750Z f='/Users/runner/work/supercollider/supercollider/build/Install/SuperCollider/SuperCollider.app/Contents/MacOS/sclang'
2024-06-13T16:22:29.2900430Z external_prereqs='/usr/local/Cellar/glib/2.80.3/lib/libglib-2.0.0.dylib'
2024-06-13T16:22:29.2901410Z ;external prerequisites found:
2024-06-13T16:22:29.2904560Z f='/Users/runner/work/supercollider/supercollider/build/Install/SuperCollider/SuperCollider.app/Contents/Frameworks/QtWebEngineCore.framework/Versions/5/Helpers/QtWebEngineProcess.app/Contents/MacOS/QtWebEngineProcess'
2024-06-13T16:22:29.2908620Z external_prereqs='/usr/local/Cellar/glib/2.80.3/lib/libglib-2.0.0.dylib'
2024-06-13T16:22:29.2910120Z '
2024-06-13T16:22:29.2910560Z -- 
2024-06-13T16:22:29.2911650Z CMake Error at /usr/local/Cellar/cmake/3.29.3/share/cmake/Modules/BundleUtilities.cmake:1127 (message):
2024-06-13T16:22:29.2913800Z   error: verify_app failed
2024-06-13T16:22:29.2915160Z Call Stack (most recent call first):
2024-06-13T16:22:29.2916030Z   editors/sc-ide/cmake_install.cmake:110 (verify_app)
2024-06-13T16:22:29.2916880Z   editors/cmake_install.cmake:44 (include)
2024-06-13T16:22:29.2917710Z   cmake_install.cmake:63 (include)
2024-06-13T16:22:29.2918540Z 
2024-06-13T16:22:29.2918550Z 
2024-06-13T16:22:29.2996080Z 
2024-06-13T16:22:29.4947040Z ** BUILD FAILED **
```

See:
* https://formulae.brew.sh/formula/glib (which got recently updated? https://github.com/Homebrew/homebrew-core/blob/8c1cad8e067d70947ab610265030aaa9a6cd4b37/Formula/g/glib.rb )
* https://formulae.brew.sh/formula/cmake#default

I don't know if it is a problem on their end or on ours, but this will break all builds currently, so it is best to be aware of this.

## Environment

* SuperCollider version: macOS Universal in CI
* Operating system: macOS
* Other details (Qt version, audio driver, etc.): universal build - signing fails

## Steps to reproduce

run CI

## Expected vs. actual behavior

CI should pass

**Comment by JordanHendersonMusic** [comment]
Any way of setting the ci up to optionally skip some steps so the SC test can still run?

**Comment by capital-G** [comment]
> Any way of setting the ci up to optionally skip some steps so the SC test can still run?

I think this is what #6340 is about, but I think it is not as easy as one may think, see

https://github.com/supercollider/supercollider/blob/d656772bd22c2c402c2cdde90db31b37a13dbbac/.github/workflows/actions.yml#L589

**Comment by capital-G** [comment]
So I took a look at this, it seems a problem could be this script

https://github.com/supercollider/supercollider/blob/0951aea91f40781d139dd77daa3ba63a4d55afab/.github/workflows/actions.yml#L308-L315

https://gist.githubusercontent.com/dyfer/6c83905d4593750105897e51e87ec345/raw/1f89f2fe767a2e637387e74f0c8dcc86a0963f60/brew-install-universal.sh

It seems brew does not offer universal libraries, that is the reason why the script is necessary.

Maybe @dyfer can give a hint if there is an easy way to fix it? :)

**Comment by silvanocerza** [comment]
Separate Arm and x86 builds are better in my opinion. The binaries created are much smaller.

Not sure if the build process would be easier though. I can give it a try as I have an M1 Pro. 🤷

**Comment by capital-G** [comment]
I inspected this further and the problem seems to be originating during the qt installation?

```log
2024-06-13T16:12:10.2714190Z [1;37mpackage: [1;32mglib[0m
2024-06-13T16:12:10.2741920Z [1;37mglib: downloading native package[0m
2024-06-13T16:12:13.4739490Z file sha 00a8c3e38de4d37541797d7adc1a0859cd3d357f4b5757d8828dc4c105ebb32f doesn't match sha from 'brew info' b6761564a1021728437898d4b91c5468cf237ec972a06abc296082ca5c32acd1
2024-06-13T16:12:14.9691300Z downloading glib.monterey.bottle.tar.gz from https://ghcr.io/v2/homebrew/core/glib/blobs/sha256:b6761564a1021728437898d4b91c5468cf237ec972a06abc296082ca5c32acd1
2024-06-13T16:12:14.9855630Z   % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
2024-06-13T16:12:14.9856830Z                                  Dload  Upload   Total   Spent    Left  Speed
2024-06-13T16:12:14.9858320Z 
2024-06-13T16:12:15.2474400Z   0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
2024-06-13T16:12:15.2476040Z   0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
2024-06-13T16:12:15.4924120Z 
2024-06-13T16:12:15.6103770Z  53 8452k   53 4540k    0     0  9023k      0 --:--:-- --:--:-- --:--:-- 9023k
2024-06-13T16:12:15.6139740Z [1;37mglib: processing native package[0m
2024-06-13T16:12:15.7034990Z 100 8452k  100 8452k    0     0  13.2M      0 --:--:-- --:--:-- --:--:-- 31.5M
2024-06-13T16:12:15.7051310Z decompressing glib.monterey.bottle.tar.gz
2024-06-13T16:12:19.6356220Z 2024-06-13 16:12:19.634 xcodebuild[19511:47278] Requested but did not find extension point with identifier Xcode.IDEKit.ExtensionSentinelHostApplications for extension Xcode.DebuggerFoundation.AppExtensionHosts.watchOS of plug-in com.apple.dt.IDEWatchSupportCore
2024-06-13T16:12:19.6361560Z 2024-06-13 16:12:19.635 xcodebuild[19511:47278] Requested but did not find extension point with identifier Xcode.IDEKit.ExtensionPointIdentifierToBundleIdentifier for extension Xcode.DebuggerFoundation.AppExtensionToBundleIdentifierMap.watchOS of plug-in com.apple.dt.IDEWatchSupportCore
2024-06-13T16:12:22.0324320Z --- /Users/runner/work/supercollider/supercollider/brew-universal/extracted_native/glib/2.80.3/bin/gtester ---
2024-06-13T16:12:22.0395480Z @@HOMEBREW_CELLAR@@/glib/2.80.3/lib/libglib-2.0.0.dylib: changing path to /usr/local/Cellar/glib/2.80.3/lib/libglib-2.0.0.dylib
2024-06-13T16:12:22.9708260Z --- /Users/runner/work/supercollider/supercollider/brew-universal/extracted_native/glib/2.80.3/bin/gsettings ---
2024-06-13T16:12:22.9798220Z @@HOMEBREW_CELLAR@@/glib/2.80.3/lib/libgio-2.0.0.dylib: changing path to /usr/local/Cellar/glib/2.80.3/lib/libgio-2.0.0.dylib
2024-06-13T16:12:23.0018320Z @@HOMEBREW_CELLAR@@/glib/2.80.3/lib/libglib-2.0.0.dylib: changing path to /usr/local/Cellar/glib/2.80.3/lib/libglib-2.0.0.dylib
```

But maybe the verifying is the actual issue, as otherwise the building would fail? I am completly clueless in the multi-platform-finesse of macOS builds and homebrew.
I also found this issue https://gitlab.com/kicad/code/kicad/-/issues/15376

**Comment by capital-G** [comment]
#6358 introduces a temporary fix

**Comment by capital-G** [comment]
A bit of an update - I tried to add an additional native macOS arm target based on #6334 and it isn't too hard, see https://github.com/capital-G/supercollider/actions/runs/9569427310/job/26381980023 - this would allow our build pipeline to become a bit easier and therefore more robust.
Can someone with an apple silicon give this a go? https://github.com/capital-G/supercollider/actions/runs/9569427310/artifacts/1613905335

I also took a look at our most recent dev build, i can run it when right clicking on it and click "please open it", so maybe the verify step isn't too critical after all? I think the "malicious software" is displayed because we haven't signed it with a certificate.

We could additionally try to replace the homebrew-universal hack with vcpkg or macports which seem to support universal binaries natively? It is hard to debug because I don't want to taint my own homebrew :/

BTW, it seems [macOS14+ GitHub runners only support arm runners](https://docs.github.com/en/actions/using-github-hosted-runners/about-github-hosted-runners/about-github-hosted-runners#standard-github-hosted-runners-for-public-repositories), this could generate problems for x86 builds in the future?

Additionally, macOS11 runner seems to be offline starting in 10 days 🚒  happy times ahead.

![image](https://github.com/supercollider/supercollider/assets/8267062/6678339c-3f4b-4bde-b865-083e59b15bec)

**Comment by dyfer** [comment]
Thanks for this!

Yes, my script for making universal binaries is indeed fragile. But I'm not sure if the failure was because of it... 
The interaction between macdeployqt and library paths in homebrew is a source of previous problems in that regard.
At the very least the build still seems to work, without the verification (confirmed on macOS M1).

Finally, I don't oppose splitting the builds to x64 and arm, though that's one more download link to maintain. The "homebrew universal" script was done at the time when there were no arm runners available.

**Comment by capital-G** [comment]
@dyfer can you give a quick sketch what the verify_app step actually does? I looked at [CMake documentation](https://cmake.org/cmake/help/latest/module/BundleUtilities.html) but I am as enlightened as before reading it

> Verifies that an application <app> appears valid based on running analysis tools on it. Calls [message(FATAL_ERROR)](https://cmake.org/cmake/help/latest/command/message.html#command:message) if the application is not verified.

I think of giving the CI a rewrite by replacing the matrix with dedicated build steps. I think this would make tracking of what actually happens a bit easier and removes barriers such as that macOS testing requires on linux builds - excessive duplication can be outsourced into a script.

It seems an alternative way of obtaining universal dependencies would be by using macports which, contrary to homebrew, support universal builds - but one needs to compile everything by oneself, so it is a bit of a Gentoo style - don't know if such a thing is really feasible, maybe with enough caching it is.

**Comment by silvanocerza** [comment]
Somewhat related to this but I couldn't find a thourough guide on building the project on different platforms and architecture.

I kinda tried to follow the CI steps to build locally but I met some barriers on macOS M1, so I'm not sure how to proceed. 

Ideally I would like to have just a couple of commands to configure and build locally.

**Comment by dyfer** [comment]
> @dyfer can you give a quick sketch what the verify_app step actually does? I looked at [CMake documentation](https://cmake.org/cmake/help/latest/module/BundleUtilities.html) but I am as enlightened as before reading it

Hahaha yeah I'm certainly not an expert on this.
From memory:
I _think_ it looks at all loader_paths and similar stuff in all the binaries (and libraries) in the bundle and checks whether they point to locations inside the bundle.

`macdeployqt` is the actual tool that we use to assemble the app on macOS and that's supposed to change all the library paths (loader_paths etc) inside the binaries to point to local (bundle) library locations (as the libraries like qt and libsndfile etc are copied into the bundle). Curiously, verify_app doesn't consider locations that use @rpath as valid but that's what the official Qt libraries come configured with. So verify_app complains on bundle made by macdeployqt from the official qt install, even though the bundle works fine... (I haven't thought about this in months so I might've gotten some details wrong, but I think this is/was the case)

> I think of giving the CI a rewrite by replacing the matrix with dedicated build steps after #6354 gets merged. I think this, combined with a [CMakePresets.json](https://github.com/supercollider/supercollider/pull/6351/files#diff-1712b235e02bd4a3ad866fe26447a6dc5c1b449faeea513f449d244d46d03cf5) introduced in #6351, would make tracking of what actually happens a bit easier and removes barriers such as that macOS testing requires on linux builds - excessive duplication can be outsourced into a script.

I would love having the tests run independently of failures on other platforms.
When I looked at making that happen in GHA yaml, my feeling was that it seemed that it's easy to make the CI configuration too fragmented IMO, which makes it harder to maintain unless you are really invested in it. However, maybe that's a totally unwarranted opinion and we should go that route.

> It seems an alternative way of obtaining universal dependencies would be by using macports which, contrary to homebrew, support universal builds - but one needs to compile everything by oneself, so it is a bit of a Gentoo style - don't know if such a thing is really feasible, maybe with enough caching it is... ¯\_(ツ)_/¯

Ha, using `macports` definitely did cross my mind when I was working on the macOS universal build.
However, maybe let's not add _yet another_ package manager docs/configuration...
I went the `vcpkg` route. The CI and CMake is already set up to support it. However, the ability of VCPKG to build universal binaries is spotty... i.e. I had problems making universal libsndfile (one of the mpeg dependencies) build on x64 runners (but I was able to use vcpkg to build that library locally on arm macOS as universal binary).

What I'm saying is that I think that official Qt6 (which is universal binary) - once we migrate - plus small misc libs from vcpkg might be the most straightforward path for now.
We probably still should support homebrew for local builds though...
